Angular i18 translation extractor

jebner opened the pull request.

Commit:
add support for angular i18n

- new extractor for extracting translations and ids from i18n and i18n-*
attributes

Commit:
export new translation extractor

- add test for using it in html-loader's options

Commit:
add section in readme for custom html extractors

angheladrian commented on src/html/angular-i18n-translations-extractor.ts:
```diff
@@ -0,0 +1,83 @@
+const I18N_ATTRIBUTE_NAME = "i18n";
+const ID_INDICATOR = "@@";
+
+export default function angularI18nTranslationsExtractor(
```
A jsdoc for the `angularI18nTranslationsExtractor` would be nice: why is the purpose more exactly.

angheladrian commented on src/html/angular-i18n-translations-extractor.ts:
```diff
@@ -0,0 +1,83 @@
+): void {
+    const translationId = extractTranslationId(attribute, context);
+
+    let defaultTranslation: string;
```
the `defaultTranslation` is needed only inside the `if` so it can be move there.
The `let` could be replaced with a `const`, or even better remove the `defaultTranslation` and use directly `element.texts[0].text`

angheladrian commented on src/html/angular-i18n-translations-extractor.ts:
```diff
@@ -0,0 +1,83 @@
+            position: element.startPosition
+        });
+    } else {
+        context.emitError(`The element ${context.asHtml()} with attribute  ${attribute.name} is empty and is therefore missing the default translation.`, attribute.startPosition);
```
The error will be emitted also when there is no `translationId`. For that case the error message would miss-leading.

angheladrian commented on src/html/angular-i18n-translations-extractor.ts:
```diff
@@ -0,0 +1,83 @@
+): void {
+
+    for (const i18nAttribute of i18nAttributes) {
+        const translationId = extractTranslationId(i18nAttribute, context);
```
I suggest extracting the for block content into a separated function. This would allow a quicker return when the  are no `translationId`, `attributeName` or `defaultText`. In this way the some of the checks can be removed.

angheladrian commented on README.MD:
```diff
@@ -74,6 +74,74 @@ The loaders accepts the name of a loader that should be applied in advance. E.g.
 
+### Custom HTML Translation extractors
+
+The htmlLoader supports registering custom HTML text extractors and rewrites the AngularJS extractor to be based on the same API.
```
I find this sentence a little confusing. Plus is the _AngularJS extractor_ rewritten, or an extra one is added.

angheladrian commented on README.MD:
````diff
@@ -74,6 +74,74 @@ The loaders accepts the name of a loader that should be applied in advance. E.g.
+}
+```
+
+WebpackAngularTranslates provides the `angularI18nTranslationsExtractor` to support extractions of translations in hybrid applications 
````
I suggest adding a new section here for `angularI18nTranslationsExtractor`

angheladrian replied:
Plus is this extractor only for hybrid applications, or it can be used for angular only also?

Commit:
improve docs about translation extractors

Commit:
code improvements

- do not assign unnecessary to defaultTranslation variable
- make sure missing default translation error message is only fired for
the correct case
- extract for loop block into separate function

angheladrian committed:
Update angular-i18n-translations-extractor.ts

angheladrian commented on src/html/angular-i18n-translations-extractor.ts:
```diff
@@ -27,15 +43,13 @@ function handleTranslationsOfElements(
             position: element.startPosition
         });
-    } else {
+    } else if (translationId) {
```
Probably should be `!translationId`. But I would actually check for `element.texts.length` since this is what the message is about.

Commit:
check for missing texts instead of translationId
